File: packages/llm-test-harness/src/adapters/openai.ts
import type { CassetteInteraction, CassetteRequest, ResolvedHarnessConfig } from '../types.js'
import type { CassetteStore } from '../cassette/cassette-store.js'
import { hashRequest } from '../cassette/cassette-hash.js'
import { findInteraction } from '../cassette/cassette-match.js'
import { CassetteMissError, CassetteOverwriteError } from '../errors.js'

// Minimal type definitions for the OpenAI SDK shapes.

interface OpenAIMessage {
  role: 'system' | 'user' | 'assistant' | 'tool'
  content: string | null
  tool_calls?: Array<{
    id: string
    type: 'function'
    function: { name: string; arguments: string }
  }>
}

interface OpenAITool {
  type: 'function'
  function: {
    name: string
    description?: string
    parameters?: Record<string, unknown>
  }
}

interface OpenAICreateParams {
  model: string
  messages: OpenAIMessage[]
  max_tokens?: number | null
  temperature?: number | null
  top_p?: number | null
  stop?: string | string[] | null
  stream?: boolean
  tools?: OpenAITool[]
  [key: string]: unknown
}

interface OpenAIChatCompletion {
  id: string
  object: 'chat.completion'
  created: number
  model: string
  choices: Array<{
    index: number
    message: { role: 'assistant'; content: string | null; tool_calls?: unknown[] }
    finish_reason: string | null
    logprobs: null
  }>
  usage: { prompt_tokens: number; completion_tokens: number; total_tokens: number }
}

function normalizeRequest(params: OpenAICreateParams): CassetteRequest {
  // Extract system message from messages array (OpenAI embeds it as role='system')
  const systemMsg = params.messages.find((m) => m.role === 'system')
  const nonSystemMsgs = params.messages.filter((m) => m.role !== 'system')

  return {
    provider: 'openai',
    model: params.model,
    system: systemMsg?.content ?? null,
    messages: nonSystemMsgs
      .filter((m) => m.role === 'user' || m.role === 'assistant')
      .map((m) => ({ role: m.role as 'user' | 'assistant', content: m.content ?? '' })),
    params: {
      max_tokens: params.max_tokens ?? null,
      temperature: params.temperature ?? null,
      top_p: params.top_p ?? null,
      stop: params.stop ?? null,
    },
    tools: params.tools
      ? params.tools.map((t) => ({
          name: t.function.name,
          description: t.function.description ?? null,
        }))
      : null,
  }
}

function buildOpenAIResponse(interaction: CassetteInteraction): OpenAIChatCompletion {
  const { response, metadata } = interaction
  const textContent = response.content?.[0]?.text ?? ''

  return {
    id: metadata.provider_request_id ?? `replayed_${interaction.id.slice(0, 16)}`,
    object: 'chat.completion',
    created: Math.floor(new Date(metadata.recorded_at).getTime() / 1000),
    model: interaction.request.model,
    choices: [
      {
        index: 0,
        message: { role: 'assistant', content: textContent },
        finish_reason: response.finish_reason ?? response.stop_reason ?? 'stop',
        logprobs: null,
      },
    ],
    usage: {
      prompt_tokens: response.usage?.input_tokens ?? 0,
      completion_tokens: response.usage?.output_tokens ?? 0,
      total_tokens:
        response.usage?.total_tokens ??
        (response.usage?.input_tokens ?? 0) + (response.usage?.output_tokens ?? 0),
    },
  }
}

async function interceptCreate(
  params: OpenAICreateParams,
  originalFn: (params: OpenAICreateParams) => Promise<OpenAIChatCompletion>,
  store: CassetteStore,
  config: ResolvedHarnessConfig,
): Promise<OpenAIChatCompletion> {
  if (params.stream === true) {
    return originalFn(params)
  }

  const req = normalizeRequest(params)
  const hash = hashRequest(req)

  if (config.mode === 'replay' || config.mode === 'hybrid') {
    const interaction = findInteraction(store, req)
    if (interaction) {
      return buildOpenAIResponse(interaction)
    }
    if (config.mode === 'replay') {
      throw new CassetteMissError(hash)
    }
  }

  if (config.mode === 'passthrough') {
    return originalFn(params)
  }

  const start = Date.now()
  const response = await originalFn(params)
  const durationMs = Date.now() - start

  const firstChoice = response.choices[0]
  const interaction: CassetteInteraction = {
    id: hash,
    request: req,
    response: {
      type: 'message',
      content: firstChoice?.message.content
        ? [{ type: 'text', text: firstChoice.message.content }]
        : [],
      usage: {
        input_tokens: response.usage.prompt_tokens,
        output_tokens: response.usage.completion_tokens,
        total_tokens: response.usage.total_tokens,
      },
      finish_reason: firstChoice?.finish_reason ?? undefined,
    },
    metadata: {
      recorded_at: new Date().toISOString(),
      duration_ms: durationMs,
      provider_request_id: response.id,
    },
  }

  const final = config.onBeforeRecord(interaction)

  if (config.noOverwrite && store.findById(hash) !== undefined) {
    throw new CassetteOverwriteError(hash)
  }

  store.append(final)
  return response
}

export function wrapOpenAIClient<T extends object>(
  client: T,
  store: CassetteStore,
  config: ResolvedHarnessConfig,
): T {
  return new Proxy(client, {
    get(target, prop, receiver) {
      if (prop === 'chat') {
        const originalChat = Reflect.get(target, prop, target) as Record<string, unknown>
        return new Proxy(originalChat, {
          get(chatTarget, chatProp) {
            if (chatProp === 'completions') {
              const originalCompletions = Reflect.get(chatTarget, chatProp, chatTarget) as Record<
                string,
                unknown
              >
              return new Proxy(originalCompletions, {
                get(compTarget, compProp) {
                  if (compProp === 'create') {
                    return (params: OpenAICreateParams) =>
                      interceptCreate(
                        params,
                        (p) =>
                          (
                            compTarget['create'] as (
                              p: OpenAICreateParams,
                            ) => Promise<OpenAIChatCompletion>
                          ).call(compTarget, p),
                        store,
                        config,
                      )
                  }
                  return Reflect.get(compTarget, compProp, compTarget)
                },
              })
            }
            return Reflect.get(chatTarget, chatProp, chatTarget)
          },
        })
      }
      return Reflect.get(target, prop, receiver)
    },
  })
}

File: packages/llm-test-harness-python/llm_test_harness/adapters/openai.py
from __future__ import annotations

import time
from datetime import datetime, timezone
from typing import Any

from ..cassette.match import find_interaction, require_interaction
from ..cassette.store import CassetteStore
from ..errors import CassetteOverwriteError
from ..types import (
    CassetteContentBlock,
    CassetteInteraction,
    CassetteMessage,
    CassetteMetadata,
    CassetteParams,
    CassetteRequest,
    CassetteResponse,
    CassetteUsage,
    ResolvedHarnessConfig,
)


def _build_request(params: dict[str, Any]) -> CassetteRequest:
    raw_messages = params.get("messages", [])
    system = None
    messages: list[CassetteMessage] = []
    for m in raw_messages:
        if m.get("role") == "system" and system is None:
            system = m.get("content")
        else:
            messages.append(CassetteMessage(role=m["role"], content=m["content"]))

    tools_raw = params.get("tools")
    tools = None
    if tools_raw:
        from ..types import CassetteTool
        tools = []
        for t in tools_raw:
            fn = t.get("function", {}) if t.get("type") == "function" else t
            tools.append(CassetteTool(name=fn.get("name", ""), description=fn.get("description")))

    cp = CassetteParams(
        max_tokens=params.get("max_tokens"),
        temperature=params.get("temperature"),
        top_p=params.get("top_p"),
        stop=params.get("stop"),
    )
    return CassetteRequest(
        provider="openai",
        model=params.get("model", ""),
        system=system,
        messages=messages,
        params=cp,
        tools=tools,
    )


def _response_to_cassette(resp: Any) -> CassetteResponse:
    content = None
    choices = getattr(resp, "choices", None) or []
    if choices:
        choice = choices[0]
        msg = getattr(choice, "message", None)
        if msg:
            content = []
            text = getattr(msg, "content", None)
            if text:
                content.append(CassetteContentBlock(type="text", text=text))
            tool_calls = getattr(msg, "tool_calls", None) or []
            for tc in tool_calls:
                fn = getattr(tc, "function", None)
                content.append(
                    CassetteContentBlock(
                        type="tool_use",
                        id=getattr(tc, "id", None),
                        name=getattr(fn, "name", None) if fn else None,
                        input=getattr(fn, "arguments", None) if fn else None,
                    )
                )

    usage = None
    u = getattr(resp, "usage", None)
    if u:
        usage = CassetteUsage(
            input_tokens=getattr(u, "prompt_tokens", None),
            output_tokens=getattr(u, "completion_tokens", None),
            total_tokens=getattr(u, "total_tokens", None),
        )

    finish_reason = None
    if choices:
        finish_reason = getattr(choices[0], "finish_reason", None)

    return CassetteResponse(
        type="chat.completion",
        content=content,
        usage=usage,
        finish_reason=finish_reason,
    )


def _cassette_to_response(interaction: CassetteInteraction) -> Any:
    """Reconstruct a minimal OpenAI-like ChatCompletion object from cassette data."""

    class _FunctionCall:
        def __init__(self, name: str, arguments: Any) -> None:
            self.name = name
            self.arguments = arguments

    class _ToolCall:
        def __init__(self, tc_id: str, fn: _FunctionCall) -> None:
            self.id = tc_id
            self.type = "function"
            self.function = fn

    class _Message:
        def __init__(self, content: Any, tool_calls: list[Any]) -> None:
            self.role = "assistant"
            self.content = content
            self.tool_calls = tool_calls or None

    class _Choice:
        def __init__(self, message: _Message, finish_reason: Any) -> None:
            self.index = 0
            self.message = message
            self.finish_reason = finish_reason

    class _Usage:
        def __init__(self, prompt: Any, completion: Any, total: Any) -> None:
            self.prompt_tokens = prompt
            self.completion_tokens = completion
            self.total_tokens = total

    class _ChatCompletion:
        def __init__(self, choices: list[Any], usage: Any, model: str) -> None:
            self.object = "chat.completion"
            self.model = model
            self.choices = choices
            self.usage = usage
            self.id = f"replay_{interaction.id[:16]}"

    text_content = None
    tool_calls: list[Any] = []
    if interaction.response.content:
        for b in interaction.response.content:
            if b.type == "tool_use":
                tool_calls.append(
                    _ToolCall(b.id or "", _FunctionCall(b.name or "", b.input))
                )
            elif b.type == "text":
                text_content = b.text

    msg = _Message(text_content, tool_calls)
    choice = _Choice(msg, interaction.response.finish_reason)

    usage_obj = None
    if interaction.response.usage:
        usage_obj = _Usage(
            interaction.response.usage.input_tokens,
            interaction.response.usage.output_tokens,
            interaction.response.usage.total_tokens,
        )

    return _ChatCompletion([choice], usage_obj, interaction.request.model)


def _intercept_create(
    original_create: Any,
    params: dict[str, Any],
    store: CassetteStore,
    config: ResolvedHarnessConfig,
) -> Any:
    if params.get("stream"):
        return original_create(**params)

    req = _build_request(params)
    mode = config.mode

    if mode == "replay":
        interaction = require_interaction(store, req)
        return _cassette_to_response(interaction)

    if mode == "passthrough":
        return original_create(**params)

    if mode in ("record", "hybrid"):
        if mode == "hybrid":
            existing = find_interaction(store, req)
            if existing is not None:
                return _cassette_to_response(existing)

        start = time.monotonic()
        resp = original_create(**params)
        duration_ms = int((time.monotonic() - start) * 1000)

        from ..cassette.hash import hash_request
        hash_ = hash_request(req)

        if config.no_overwrite and store.find_by_id(hash_) is not None:
            raise CassetteOverwriteError(hash_)

        metadata = CassetteMetadata(
            recorded_at=datetime.now(timezone.utc).isoformat(),
            duration_ms=duration_ms,
            provider_request_id=getattr(resp, "id", None),
        )
        interaction = CassetteInteraction(
            id=hash_,
            request=req,
            response=_response_to_cassette(resp),
            metadata=metadata,
        )
        if config.on_before_record:
            interaction = config.on_before_record(interaction)
        store.append(interaction)
        return resp

    return original_create(**params)


class _CompletionsProxy:
    def __init__(self, completions: Any, store: CassetteStore, config: ResolvedHarnessConfig) -> None:
        self._completions = completions
        self._store = store
        self._config = config

    def create(self, **params: Any) -> Any:
        return _intercept_create(self._completions.create, params, self._store, self._config)

    def __getattr__(self, name: str) -> Any:
        return getattr(self._completions, name)


class _ChatProxy:
    def __init__(self, chat: Any, store: CassetteStore, config: ResolvedHarnessConfig) -> None:
        self._chat = chat
        self._store = store
        self._config = config

    @property
    def completions(self) -> _CompletionsProxy:
        return _CompletionsProxy(self._chat.completions, self._store, self._config)

    def __getattr__(self, name: str) -> Any:
        return getattr(self._chat, name)


class OpenAIClientProxy:
    def __init__(self, client: Any, store: CassetteStore, config: ResolvedHarnessConfig) -> None:
        self._client = client
        self._store = store
        self._config = config

    @property
    def chat(self) -> _ChatProxy:
        return _ChatProxy(self._client.chat, self._store, self._config)

    def __getattr__(self, name: str) -> Any:
        return getattr(self._client, name)
